src/pmem.c

```diff
@@ -602,17 +653,243 @@ memmove_nodrain_normal(void *pmemdest, const void *src, size_t len)
 
-	/* XXX stub version */
-	return memmove(pmemdest, src, len);
+	/*
+	 * The way this works is:
+	 * 	Calculate the non-overlapping size
+	 * 	Break the memmove into these chunk sizes
+	 *	Determine if we are copying forward/backward
+	 * This handles all cases, including overlap. Consider the following:
+	 *
+	 * 	src
+	 * high addr				dest
+	 *					high addr
+	 * ------
+	 * |	|s3-----------------------------------------------V
+	 * ------						--------
+	 * |	|s2-----------------------------|		|	|d3
+	 * ------				|-------|	--------
+	 * |	|s1---------------------------- 	|------>|	|d2
+	 * ------				|		--------
+	 *					|-------------->|	|d1
+	 * **** 						--------
+	 *
+	 * In this case we copy down, starting with s1 to d1, so that we
+	 * never overwrite the source before moving it. Any leftover
+	 * bytes are handled after moving the chunk size data.
+	 * This works similarly when src < dest.
+	 */
+	if (src > pmemdest)
+		chunk_size = util_nonoverlap_range((void *)src1, dest1, len);
+	else
+		chunk_size = util_nonoverlap_range(dest1, src1, len);
+
+	/*
+	 * The most common case is non-overlapping addresses.
+	 */
+	if (chunk_size == len) {
+		pmem_align(&dest1, src1, &cnt, UP);
+		src1 += cnt;
+		len -= cnt;
+		return (pmemdest = (nodrain_movnt(dest1, src1, len)));
+	}
+
+	/* overlapping addresses, possible unaligned dest. */
+	if ((uint64_t)src1 > (uint64_t)dest1) {
+		pmem_align(&dest1, src1, &cnt, UP);
+		len -= cnt;
+		src1 += cnt;
+
+		/* Align the chunk size */
+		if (chunk_size > MOVNT_SIZE) {
+			if ((mod = (chunk_size % MOVNT_SIZE)) != 0)
+				chunk_size -= mod;
+		}
+		num_chunks = len/chunk_size;
+
+		for (i = 0; i < num_chunks; i ++) {
+			nodrain_movnt(dest1, src1, chunk_size);
+			src1 += chunk_size;
+			len -= chunk_size;
+			dest1 += chunk_size;
+		}
+
+		if (len != 0)
+			dest1 = nodrain_movnt(dest1, src1, len);
+	} else {
+
+
+		if (chunk_size	> MOVNT_SIZE) {
+			if ((mod = (chunk_size % MOVNT_SIZE)) != 0)
+				chunk_size -= mod;
+		}
+
+		dest1 = dest1 + len;
+		src1 = src1 + len;
+		pmem_align(&dest1, src1, &cnt, DOWN);
+		src1 -= cnt;
+		len -= cnt;
+
+		/* Reduce len by 1 to not exceed the mapped memory. */
+		len -= 1;
+
+		/* Start at the 1st chunk address so we can copy up */
+		dest1 -= chunk_size;
+		src1 -= chunk_size;
+
+		num_chunks = len/chunk_size;
+
+		for (i = 0; i < num_chunks - 1; i ++) {
+			nodrain_movnt(dest1, src1, chunk_size);
+			src1 -= chunk_size;
+			dest1 -= chunk_size;
+			len -= chunk_size;
+		}
+
+		if (len != 0)
+			dest1 = nodrain_movnt(dest1, src1, len);
+	}
+	return pmemdest;
+}
+
+/*
+ * memcpy_nodrain_movnt -- (internal) memcpy to pmem without hw drain, movnt
+ */
+static void *
+memcpy_nodrain_movnt(void *pmemdest, const void *src, size_t len)
+{
+
+	size_t cnt;
+	void *dest1 = pmemdest;
+
+	/* way too chatty for LOG level 3 */
+	LOG(15, "pmemdest %p src %p len %zu", pmemdest, src, len);
+
+	pmem_align(&dest1, src, &cnt, UP);
```

Comment:
IMHO, the name *pmem_align* is quite misleading. It suggests the function simply aligns the *dest* address, while it actually copies some data as well. I would rather rename it to *pmem_memcpy_unaligned_nodrain* or something like that.